# SNT - DONNÉES STRUCTURÉES - TP2 - PYTHON CSV

*Données au format CSV*

## La bibliothèque PANDAS

Pour manipuler les fichiers CSV, Python embarque une bibliothèque nommée 'csv'. Cependant, nous allons une autre bilibothèque plus performante avec les fichiers de taille importante : 'pandas'.

**Préalable**

Commençons par nous assurer que cette bibliothèque est installée dans Thonny ou Spyder ou Pyzo (etc.):
- Dans le menu *OUTILS*, choisir *GERER LES PAQUETS* ;
- chercher 'pandas' : choisir *INSTALLER* si ce n'est pas déjà le cas.

---
---
*REMARQUE*

On peut aussi utiliser la commande :  

```python
pip install pandas
```

    
dans la console, qui fonctionne aussi avec Pyzo, EduPython, etc.

---
---

Créer un NOUVEAU FICHIER dans Thonny et enregistrez-le sur votre clé USB sous le nom :
    
    numéro_binôme OU Nom élève - TP2-THEME5-CSV PYTHON

**Dans le même dossier**, vous placerez le fichier 'ClassDIffFilms20092018.csv' que vous récupérerez sur TEAMS.

## C'est parti !

### TOUTES LES INSTRUCTIONS SERONT ÉCRITES DANS L'ÉDITEUR DE CODE AFIN QUE VOUS PUISSIEZ ENREGISTRER L'ENSEMBLE DES INSTRUCTIONS MANIPULÉES, MAIS VOUS LES TESTEREZ UNE À UNE EN LES COPIANT DANS LA CONSOLE (SHELL) LES UNES APRÈS LES AUTRES :

<img src = 'img test.png' width='400'>

Commençons par importer cette bibliothèque Pandas :

In [1]:
import pandas

## Charger en mémoire un fichier de données au format 'csv' :

On remarquera deux paramètres lorsque l'on appelle la fonction :
- le nom du fichier,
- le séparateur de données (ici, la virgule)

In [2]:
monfichier = pandas.read_csv('ClassDIffFilms20092018.csv', sep = ';')

### Informations concernant la collection de données

Dans un premier temps, nous pouvons obtenir le volume de données contenu dans cette collection. On fait alors appel à la méthode 'shape' qui retourne dans l'ordre :
- le nombre d'enregistrements;
- le nombre de descripteurs.

In [3]:
print(monfichier.shape)

(133, 6)


**Remarque :
en console, on peut omettre l'instruction**
```python
print()
```
**dès qu'il s'agit d'obtenir une sortie, un affichage à l'écran :**

In [4]:
monfichier.shape

(133, 6)

### Afficher la collection de données

Nous pouvons afficher les enregistrements de notre fichier de deux façons.

Soit en invoquant le nom de notre tableau de données chargé en mémoire, ce qui fonctionnera en mode interactif (mode console ou *shell* ) :

In [5]:
monfichier


nombre de diffusions                                          titre  \
0                      19                         11 COMMANDEMENTS (LES)   
1                      18                       ASTERIX CHEZ LES BRETONS   
2                      18                   TINTIN ET LE LAC AUX REQUINS   
3                      17                                          CHAOS   
4                      17  DEUX HEURES MOINS LE QUART AVANT JESUS-CHRIST   
5                      17                  DOUZE TRAVAUX D'ASTERIX (LES)   
6                      17                 FAST AND FURIOUS : TOKYO DRIFT   
7                      16                ASTERIX ET LA SURPRISE DE CESAR   
8                      16                                       BANDIDAS   
9                      16                        BIDASSES EN FOLIE (LES)   
10                     16                                  FLIC OU VOYOU   
11                     16       LIGUE DES GENTLEMEN EXTRAORDINAIRES (LA)   
12                     16                                     P.R.O.F.S.   
13                     16                  TINTIN ET LE TEMPLE DU SOLEIL   
14                     16         VIE EST UN LONG FLEUVE TRANQUILLE (LA)   
15                     15     4 FANTASTIQUES ET LE SURFER D'ARGENT (LES)   
16                     15                           ASTERIX ET CLEOPATRE   
17                     15                   ASTERIX ET LE COUP DU MENHIR   
18                     15                         ASTERIX ET LES INDIENS   
19                     15                             ASTERIX LE GAULOIS   
20                     15                        CHATEAU DE MA MERE (LE)   
21                     15                             COMME T'Y ES BELLE   
22                     15                    KING RISING : AU NOM DU ROI   
23                     15                                         NIKITA   
24                     15                                         SAHARA   
25                     15                                 SCOUT TOUJOURS   
26                     15               TOUR MONTPARNASSE INFERNALE (LA)   
27                     15                                V POUR VENDETTA   
28                     15                                       YAMAKASI   
29                     14                                         3 AMIS   
..                    ...                                            ...   
103                    12                 IL ETAIT UNE FOIS DANS L'OUEST   
104                    12                                  INCONTROLABLE   
105                    12                           INCROYABLE HULK (L')   
106                    12                                             JO   
107                    12                                JURASSIC PARK 3   
108                    12                      JUST MARRIED (OU PRESQUE)   
109                    12                                 MEDAILLON (LE)   
110                    12                  MEILLEURS AMIS DU MONDE (LES)   
111                    12                      MEMOIRE DANS LA PEAU (LA)   
112                    12                                 MEN IN BLACK 2   
113                    12                               MONDE PERDU (LE)   
114                    12                         MORT DANS LA PEAU (LA)   
115                    12                                          OSCAR   
116                    12                POUR 100 BRIQUES T'AS PLUS RIEN   
117                    12                                  REGRETS (LES)   
118                    12                           RETOUR VERS LE FUTUR   
119                    12                         RETOUR VERS LE FUTUR 2   
120                    12                         RETOUR VERS LE FUTUR 3   
121                    12                                ROI ARTHUR (LE)   
122                    12                                     SAINT (LE)   
123                    12                            

Soit en utilisant la fonction
```python
print()```
dans votre éditeur de code. A vous de jouer !

In [6]:
print(monfichier)

     nombre de diffusions                                          titre  \
0                      19                         11 COMMANDEMENTS (LES)   
1                      18                       ASTERIX CHEZ LES BRETONS   
2                      18                   TINTIN ET LE LAC AUX REQUINS   
3                      17                                          CHAOS   
4                      17  DEUX HEURES MOINS LE QUART AVANT JESUS-CHRIST   
5                      17                  DOUZE TRAVAUX D'ASTERIX (LES)   
6                      17                 FAST AND FURIOUS : TOKYO DRIFT   
7                      16                ASTERIX ET LA SURPRISE DE CESAR   
8                      16                                       BANDIDAS   
9                      16                        BIDASSES EN FOLIE (LES)   
10                     16                                  FLIC OU VOYOU   
11                     16       LIGUE DES GENTLEMEN EXTRAORDINAIRES (LA)   
12          

## Jouons avec les données

Vous remarquerez qu'on ne voit pas grand-chose.... Par défaut, pandas n'affiche qu'un certain nombre d'enregistrements et de colonnes.  
Changeons cela immédiatement !

In [7]:
pandas.options.display.max_rows = 10
pandas.options.display.max_columns = 6

### Les descripteurs

Nous savons déterminer le nombre de descripteurs dans un fichier csv, mais il serait intéressant alors de connaître ces
descripteurs. Allons-y ! Mais n'oubliez pas : toutes ces instructions ont pour rôle d'afficher des données à l'écran donc en console, la fonction *print()* n'est pas nécessaire.

In [8]:
print(monfichier.columns)

Index(['nombre de diffusions', 'titre', 'réalisateurs', 'année de production',
       'nationalité', 'année de dernière diffusion'],
      dtype='object')


On peut très bien, maintenant que nous connaissons les descripteurs, avoir envie ou besoin d'obtenir uniquement les valeurs d'une colonne en particulier...

In [9]:
print(monfichier[['réalisateurs']])

                            réalisateurs
0    DESAGNAT François / SORRIAUX Thomas
1                    VAN LAMSWEERDE Pino
2                        LEBLANC Raymond
3                            GIGLIO Tony
4                             YANNE Jean
..                                   ...
128                           BOYER Jean
129                        LARTIGAU Eric
130                          OURY Gérard
131                         SINGER Bryan
132                         TAMAHORI Lee

[133 rows x 1 columns]


Copiez  cette instruction pour la tester.

On peut aussi afficher les valeurs de plusieurs colonnes choisies !

In [10]:
print(monfichier[['réalisateurs','nationalité']])

                            réalisateurs nationalité
0    DESAGNAT François / SORRIAUX Thomas      France
1                    VAN LAMSWEERDE Pino      France
2                        LEBLANC Raymond      France
3                            GIGLIO Tony      Europe
4                             YANNE Jean      France
..                                   ...         ...
128                           BOYER Jean      France
129                        LARTIGAU Eric      France
130                          OURY Gérard      France
131                         SINGER Bryan  Etats-Unis
132                         TAMAHORI Lee  Etats-Unis

[133 rows x 2 columns]


---
**QUESTIONS A**

---
1. QUELLE INSTRUCTION PERMET DE DÉTERMINER LE NOMBRE D'ENREGISTREMENTS ?
2. COMBIEN Y-A-T'IL D'ENREGISTREMENTS DANS CETTE COLLECTION ?
3. QUELLE INSTRUCTION PERMET D'OBTENIR UNIQUEMENT LES COLONNES CONTENANT LE NOM DES FILMS ET LEUR ANNÉE DE PRODUCTION ? VOUS FEREZ UNE CAPTURE D'ÉCRAN DU RÉSULTAT DE CETTE INSTRUCTION ET VOUS LA COLLEREZ.
---

### Nombre de lignes...

In [11]:
monfichier.index

RangeIndex(start=0, stop=133, step=1)

### Afficher des lignes par leur index###

Affichons les enregistrements 5, 15 et 65 :

In [12]:
monfichier.iloc[[5,15,65]]

nombre de diffusions                                       titre  \
5                     17               DOUZE TRAVAUX D'ASTERIX (LES)   
15                    15  4 FANTASTIQUES ET LE SURFER D'ARGENT (LES)   
65                    13                                  INCASSABLE   

                     réalisateurs  année de production nationalité  \
5   GOSCINNY René / UDERZO Albert                 1975      France   
15                      STORY Tim                 2007  Etats-Unis   
65          NIGHT SHYAMALAN Manoj                 2000  Etats-Unis   

    année de dernière diffusion  
5                          2018  
15                         2018  
65                         2018

Tentez d'affichez les enregistrements allant de 13 jusque 20.
    
*Aide : quelle symbole utilise-t-on dans une cellule d'un tableur signifiant 'jusque' ?*

---
**QUESTIONS B**

---
1. QUELLE INSTRUCTION PERMET D'AFFICHER LES ENREGISTREMENTS DU 50ÈME AU 65ÈME ? CAPTUREZ LE RÉSULTAT DE CETTE INSTRUCTION ET COLLEZ CETTE CAPTURE.
2. QUE RENVOIT L'INSTRUCTION 'monfichier.iloc[120,:]' ?
---

## Effectuer des recherches

L'intérêt d'utiliser Python pour manipuler des données est que, d'une part, ce dernier a de nombreux outils performants pour effectuer des recherches et manipuler du texte, et d'autre part que nous pouvons automatiser ces manipulations.

Par exemple, on peut envisager la création d'un algorithme qui détecterait des doublons dans une playlist et les supprimerait.

### Récupérer une valeur dans un enregistrement dont l'index est connu###

Il faut avoir en tête les descripteurs... ou utiliser la fonction qui nous permet de les récupérer.

---

**QUESTION C**

---
1. RAPPELEZ L'INSTRUCTION PERMETTANT D'OBTENIR LA LISTE DES DESCRIPTEURS DE LA COLLECTION. COLLEZ ICI LE RÉSULTAT DE CETTE INSTRUCTION.

---

Copiez le code suivant dans la console.

In [13]:
recherche = monfichier.loc[35,'titre']
print("Le nom du film en 35ème position dans cette liste est :", format(recherche))

Le nom du film en 35ème position dans cette liste est : CROCODILE DUNDEE 2


---
**QUESTION D**

---

1. NOTEZ L'INSTRUCTION PERMETTANT D'OBTENIR LA DERNIÈRE ANNÉE DE DIFFUSION DU 103 ÈME ENREGISTREMENT. COLLEZ UNE CAPTURE DU RÉSULTAT.

---

### Récupérer les enregistrements vérifiant une condition donnée :###

On peut vouloir connaître tous les enregistrements de la collection dont le réalisateur est 'ZEMECKIS Robert' :

In [14]:
recherche2 = monfichier.loc[(monfichier['réalisateurs'] == 'ZEMECKIS Robert'),:]
print(recherche2)

     nombre de diffusions                   titre     réalisateurs  \
96                     12            FORREST GUMP  ZEMECKIS Robert   
118                    12    RETOUR VERS LE FUTUR  ZEMECKIS Robert   
119                    12  RETOUR VERS LE FUTUR 2  ZEMECKIS Robert   
120                    12  RETOUR VERS LE FUTUR 3  ZEMECKIS Robert   

     année de production nationalité  année de dernière diffusion  
96                  1993  Etats-Unis                         2018  
118                 1985  Etats-Unis                         2017  
119                 1989  Etats-Unis                         2017  
120                 1990  Etats-Unis                         2017  


Ou alors, pour une présentation plus élégante (mais uniquement en console) : 'recherche2'.

---
**QUESTIONS E**

---
1. QUELLE EST L'INSTRUCTION PERMETTANT D'OBTENIR LA LISTE DES FILMS AMÉRICAINS ? COLLEZ UNE CAPTURE DU RÉSULTAT.
2. RAPPELEZ L'INSTRUCTION PERMETTANT DE DONNER LE NOMBRE D'ENREGISTREMENTS ET DE DESCRIPTEURS D'UN FICHIER CSV.
---

### Trier les enregistrements... ou les recherches !

Trier les enregistrements selon les valeurs d'un descripteur est tout à fait possible :

In [ ]:
trierparnation = monfichier.sort_values(by = "nationalité")
print(trierparnation)

*REMARQUE :*

On peut combiner les critères de recherche ( 'critère 1' & 'critère 2' & etc.).  

---
*Exemples à faire pâlir un corbeau noir sur son arbre, perché :*

---

```python
# on trie notre collection par titre et par nombre de diffusions, en utilisant l'opérateur logique 'and' :
print(monfichier.sort_values(by = 'titre' and 'nombre de diffusions'))

# on recherche les films du réalisateur 'ZEMECKIS ROBERT' diffusés pour la dernière fois en 2018 en utilisant l'opérateur logique & :
print(monfichier.loc[(monfichier['réalisateurs'] == 'ZEMECKIS Robert' & monfichier['année de dernière diffusion' == '2018']])
```

---
**QUESTIONS F**

---
1. NOTEZ L'INSTRUCTION PERMETTANT DE TRIER LA COLLECTION PAR ANNÉE DE PRODUCTION. COLLEZ UNE CAPTURE DU RÉSULTAT.
2. NOTEZ L'INSTRUCTION PERMETTANT DE TRIER LA COLLECTION PAR TITRE ET ANNÉE DE DIFFUSION. COLLEZ UNE CAPTURE DU RÉSULTAT.
---

## Ajouter un enregistrement !

Bien entendu, il est encore plus intéressant d'ajouter des enregistrements à notre collection s'il fallait la mettre à jour.  
Ajoutons un enregistrement à cette collection :

### Définissions notre enregistrementt

In [ ]:
nouveautitre = ({
    'nombre de diffusions' : '99',
    'titre' : 'Hack en SNT',
    'réalisateurs' : 'JOUIN A / BUNIET A',
    'année de production' : '2019',
    'nationalité' : 'Inconnue',
    'année de dernière diffusion': '2020'}
)

Affichons notre futur enregistrement pour voir si tout est correct :

In [ ]:
print(nouveautitre)

### Reste à ajouter celle enregistrement à notre collection avec la méthode `.append()`

In [ ]:
monfichier = monfichier.append(nouveautitre,ignore_index=True)

Le paramètre :`ignore_index = True` précise qu'il ne faut pas enregistrer le numéro d'index dans le fichier.

**TESTEZ CES INSTRUCTIONS PUIS AFFICHEZ LA COLLECTION COMPLÈTE**

---
**QUESTION G**

---

1. EN SUIVANT L'EXEMPLE PRÉCÉDENT, AJOUTEZ VOTRE PROPRE FILM. VOUS JOINDREZ UNE CAPTURE DES INSTRUCTIONS ET DE LA PRÉSENCE DANS LA LISTE DE CE FILM GRÂCE À L'INSTRUCTION :  

`monfichier.tail(2)`        qui affiche les deux derniers enregistrements !

---

### Et on enregistre tout cela dans un nouveau fichier en choisissant le point-virgule pour séparateur ;-) :###

In [ ]:
monfichier.to_csv('Monclassement.csv',sep=';', index = False, encoding = 'utf-8')

### Remarques :  
- le paramètre `index = False` indique qu'il ne faut pas enregistrer la colonne des index dans le fichier ;
- le paramètre `encoding = 'utf-8'` indique que les caractères doivent être encodés selon la "norme" UTF8 pour éviter les soucis avec les caractères accentués et autres.

---
**QUESTION H**

---

1. Pour vérifier que l'enregistrement du fichier *'Monclassement.csv'* s'est bien déroulé, chargez ce nouveau fichier en mémoire et affichez les 5 dernières lignes de cette collection : insérez la capture d'écran dans votre document.

---